java-frontend/src/main/java/org/sonar/java/UCFGJavaVisitor.java

```diff
@@ -393,13 +402,82 @@ private void buildAssignCall(BlockBuilder blockBuilder, IdentifierGenerator idGe
     blockBuilder.assignTo(variableWithId(destination), UCFGBuilder.call(signatureFor(symbol)).withArgs(arguments.toArray(new Expression[0])), location(tree));
   }
 
+  /**
+   * Field access can be a read or a write access, depending on the context.
+   * This method should be called for field read access.
+   */
+  private void buildFieldReadAccessInvocation(BlockBuilder blockBuilder, IdentifierGenerator idGenerator, Tree tree) {
+    Expression.FieldAccess fieldAccess = buildFieldAccess(blockBuilder, idGenerator, tree);
+    if (fieldAccess != null) {
+      Expression.Variable aux = variableWithId(idGenerator.newId());
+      blockBuilder.assignTo(aux, call("__id").withArgs(fieldAccess), location(tree));
+      idGenerator.varForExpression(tree, aux.id());
+    }
+  }
+
+  private Expression.FieldAccess buildFieldAccess(BlockBuilder blockBuilder, IdentifierGenerator identifierGenerator, Tree tree) {
+    Expression.FieldAccess fieldAccess = null;
+    if (tree.is(IDENTIFIER)) {
+      fieldAccess = buildFieldAccess((IdentifierTree) tree);
+    } else if (tree.is(MEMBER_SELECT)) {
+      fieldAccess = buildFieldAccess(blockBuilder, identifierGenerator, (MemberSelectExpressionTree) tree);
+    }
+    return fieldAccess;
+  }
+
+  private Expression.FieldAccess buildFieldAccess(BlockBuilder blockBuilder, IdentifierGenerator idGenerator, MemberSelectExpressionTree memberSelectTree) {
+    ExpressionTree lhsTree = memberSelectTree.expression();
+    Symbol rhsTreeSymbol = memberSelectTree.identifier().symbol();
+    if (!rhsTreeSymbol.isVariableSymbol()) {
+      return null;
+    }
+    Expression.Variable rightSide = variableWithId(memberSelectTree.identifier().name());
+    Expression.FieldAccess fieldAccess = null;
+    if (!rhsTreeSymbol.isStatic()) {
+      // this is a recursive call that will translate any field accesses on the left
+      // we do not call lookupExpression() because that translates array access, too
+      buildFieldReadAccessInvocation(blockBuilder, idGenerator, lhsTree);
+      Expression leftSide = idGenerator.lookupExpressionFor(lhsTree);
+      if (leftSide.equals(Expression.THIS)) {
+        fieldAccess = new Expression.FieldAccess(rightSide);
+      } else if (leftSide instanceof Expression.Variable) {
+        fieldAccess = new Expression.FieldAccess((Expression.Variable) leftSide, rightSide);
+      }
+    } else if (lhsTree.is(IDENTIFIER)) {
+      IdentifierTree lhsIdentifierTree = (IdentifierTree) lhsTree;
+      fieldAccess = new Expression.FieldAccess(
+          new Expression.ClassName(lhsIdentifierTree.symbol().type().fullyQualifiedName()),
+          rightSide);
+    }
+    return fieldAccess;
+  }
+
+  private static Expression.FieldAccess buildFieldAccess(IdentifierTree identifierTree) {
+    if (!identifierTree.symbol().isVariableSymbol()
+        || identifierTree.symbol().owner().isMethodSymbol()
+        || identifierTree.name().equals("this")) {
```

Comment:
Not sure why we are excluding this case, is it going to be caught by `lookupExpressionFor` already ?

And what about this case:
```java
class A {

  abstract class B {
    void foo() {
      A.this.myField = taint();
    }

    abstract String taint();
  }

  String myField = "foo";
}
```